pkg/terra/client/gas_price_estimator.go

```diff
@@ -0,0 +1,174 @@
+	Ucad  string `json:"ucad"`
+	Uchf  string `json:"uchf"`
+	Uaud  string `json:"uaud"`
+	Usgd  string `json:"usgd"`
+	Uthb  string `json:"uthb"`
+	Usek  string `json:"usek"`
+	Unok  string `json:"unok"`
+	Udkk  string `json:"udkk"`
+	Uidr  string `json:"uidr"`
+	Uphp  string `json:"uphp"`
+	Uhkd  string `json:"uhkd"`
+}
+
+func (gpe *FCDGasPriceEstimator) request() (map[string]sdk.DecCoin, error) {
+	req, _ := http.NewRequest(http.MethodGet, gpe.fcdURL.String(), nil)
+	resp, err := gpe.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		gpe.lggr.Errorf("error reading body, err %v", gpe.fcdURL.RequestURI(), err)
+		return nil, err
+	}
+	var prices prices
+	if err := json.Unmarshal(b, &prices); err != nil {
+		gpe.lggr.Errorf("error unmarshalling, err %v", gpe.fcdURL.RequestURI(), err)
+		return nil, err
+	}
+	results := make(map[string]sdk.DecCoin)
+	v := reflect.ValueOf(prices)
+	for i := 0; i < v.NumField(); i++ {
+		name, value := strings.ToLower(v.Type().Field(i).Name), v.Field(i).String()
+		amount, err := sdk.NewDecFromStr(value)
+		if err != nil {
+			return nil, err
+		}
+		results[name] = sdk.NewDecCoinFromDec(name, amount)
+	}
+	return results, nil
+}
+
+func (gpe *FCDGasPriceEstimator) GasPrices() (map[string]sdk.DecCoin, error) {
+	return gpe.request()
+}
+
+var _ GasPricesEstimator = (*CachingGasPriceEstimator)(nil)
+
+type CachingGasPriceEstimator struct {
+	lastPrices map[string]sdk.DecCoin
+	estimator  GasPricesEstimator
+	lggr       Logger
+}
+
+func NewCachingGasPriceEstimator(estimator GasPricesEstimator, lggr Logger) *CachingGasPriceEstimator {
+	return &CachingGasPriceEstimator{estimator: estimator, lggr: lggr}
+}
+
+func (gpe *CachingGasPriceEstimator) GasPrices() (map[string]sdk.DecCoin, error) {
+	latestPrices, err := gpe.estimator.GasPrices()
+	if err != nil {
+		if gpe.lastPrices == nil {
+			return nil, errors.Errorf("unable to get gas prices and cache is empty, err %v", err)
+		}
+		gpe.lggr.Warnf("error %v getting latest prices, using cached value %v", err, gpe.lastPrices)
+		return gpe.lastPrices, nil
+	}
+	gpe.lastPrices = latestPrices
+	return latestPrices, nil
+}
+
+type ComposedGasPriceEstimator struct {
+	estimators []GasPricesEstimator
+	lggr       Logger
+}
+
+func NewMustGasPriceEstimator(estimators []GasPricesEstimator, lggr Logger) *ComposedGasPriceEstimator {
+	return &ComposedGasPriceEstimator{estimators: estimators, lggr: lggr}
+}
+
+func (gpe *ComposedGasPriceEstimator) GasPrices() map[string]sdk.DecCoin {
+	// Try each estimator in order
+	var finalError error
+	for _, estimator := range gpe.estimators {
+		latestPrices, err := estimator.GasPrices()
+		if err != nil {
+			finalError = multierr.Combine(finalError, err)
+			gpe.lggr.Warnf("error using estimator, trying next one, err %v", err)
+			continue
+		}
+		return latestPrices
+	}
+	panic(fmt.Sprintf("no estimator succeeded errs %v", finalError))
```

Comment:
IMO the end-user should attest the GPE he built is valid/working, and handle the error as he pleases. Potentially panicking if it's early enough, or reporting the error otherwise.

We could consider adding a config validation step that would run on start and config updates. That way we any misconfiguration early. We can discuss adding something like this for a future exercise, but for now, I'd suggest simply moving the panic (error handling) to Core. 

When the Terra chain is initialized, we explicitly panic there (app) instead of implicitly here (lib).

Reply:
The naming **Must**GasPriceEstimator implies the panic, I was imagining we just add a GasPriceEstimator one should we want a non-panicking version. The use case we currently have is fcd->fixed fallback for which the MustXX version makes sense imo

Reply:
The problem I see here is nothing stops me to build a `compose(fcd_1, fcd_2)` GPE, and now I have a panic liability. I understand that how it's currently used in Core, this will not happen because of the fixed fallback, but still don't love the implicit panic in the lib.

Reply:
there is config validation on the fcd url at the moment which we handle in core and I'm not sure we need it for the fixed fallback

Reply:
its a super common pattern to have a MustXX which can panic and a XX which doesn't - so I'm not sure I follow how its implicit?

Reply:
to be clear I fully support having a non-panicking version should a use case arise